enhancements/installer/custom-dns.md

```diff
@@ -0,0 +1,574 @@
+
+Does this proposal implement a behavior that's new/unique/novel? Is it poorly
+aligned with existing user expectations?  Will it be a significant maintenance
+burden?  Is it likely to be superceded by something else in the near future?
+
+[Note: Decision made in the Design Details section can affect our answer to this
+question.]
+
+## Design Details
+
+### Open Questions [optional]
+
+One of the decisions that needs to be made early in the design process is to
+determine the best way to order OpenShift's configuration of the Load Balancer
+and the customer's configuration on their custom DNS solution.
+
+Today, the Installer configures the LB first and the generated DNS name is used
+to configure DNS CNAME records. The LB DNS name cannot be predicted in advance
+because it is generated using an unique UUID while configuring LB entries for
+API and Ingress services.
+
+Here are a few options being considered as potential solutions:
+
+#### Option 1: Add a pre-install step to configure LB and provide DNS config
+
+This solution considers adding a pre-install step that configures Load Balancers
+for API and Ingress services. The DNS names generated per subnet/Availability Zone
+for API and Ingress services are output to the user in the form of DNS CNAME
+config. Customers can use this generated output to configure their DNS servers.
+
+This step could be run as an installer command like
+"openshift-install generate-custom-dns-config" while providing it with a complete
+install-config. After thsi step, when "create cluster" is called with the
+customDNS feature enabled, the validation step will have to ensure both the LB and
+DNS have been configured.
+
+An advantage to this approach is that at the end of "create cluster" the customer
+can expect to have a fully functioning cluster.
+
+This approach suggests a new way to get to a fully functional cluster in 2 parts.
+Does this constitute an "agent in the middle" install? The cluster install is not
+actively waiting for the DNS config to be completed by the customer. Another
+con is that a customer could run the "generate-custom-dns-config" utility
+but never follow through with the DNS config. This will result in dangling LB config.
+
+#### Option 1A: Generate LB and DNS config during pre-install step
+
+Another variation of this approach would be for this pre-install utility to generate
+LB DNS names but not actually configure the Load Balancers. One of the components
+of the LB DNS Name is the infra-id that is generated by the installer. If this
+pre-install utility is run to generate an infra-id, we should make sure to add it
+to the install-config so that eventual call to "create cluster" can use the same
+infra-id. 
+
+The current understanding is that the LB DNS name follows this pattern:
+<infra-id>-ext-<unknown-id>.elb.<region>.<cluster domanin name>
+
+[TODO: We are still figuring out if the "unknown id" is a randomly generated
+load-balancer-resource-id or a known subnet-id.]
```

Comment:
For every instance I've inspected the <unknown id> corresponds to the load balancer id, so we cannot know this value until the load balancer is created.

Reply:
That would effectively rule out Option 1A.